New feature: model adaptivity (#198)
precice/micro-manager: docs/model-adaptivity.md

---
title: Adaptive switching of simulation models
permalink: tooling-micro-manager-model-adaptivity.html
keywords: tooling, macro-micro, two-scale, model-adaptivity

> summary: Micro Manager can switch micro models adaptively.

Reviewer: ```suggestion
summary: Micro Manager can adaptively switch models of micro simulations.
```

---

## Main Concept

> For scenarios such as FE2 simulations computing all or perhaps only active micro simulations

Reviewer: ```suggestion
For certain multiscale scenarios, having an adaptivity strategy that groups the micro simulations into active and inactive simulations
```

> may be still computationally infeasible. The alternative here is to reduce model complexity via reduced order models (ROMs) or similar.

Reviewer: ```suggestion
may be insufficient. Alternatively, a hierarchy of micro-scale models, for example, reduced order models (ROMs) can be used.
```

> Towards this the model adaptivity feature allows for the definition of multiple

Reviewer: ```suggestion
The model adaptivity functionality allows for the definition of multiple
```

> model fidelities and the switching between those at run-time.

Reviewer: ```suggestion
model fidelities and the switching between them at run-time.
```

### Iterative Process

**Without** model adaptivity the call to `micro_sim_solve(micro_sims_input, dt)` within the micro_manager would just

> provide each micro simulation with its input, solve it and return the output.

Reviewer: ```suggestion
**Without** model adaptivity, the Micro Manager calls the `solve(micro_sims_input, dt)` routine of all active simulations
and copies their output to their closest similar inactive counterparts.
```

**With** model adaptivity this becomes an iterative process, as a model may not be sufficiently accurate (given the current input).

> Thus, the call to `micro_sim_solve(micro_sims_input, dt)` with model adaptivity results in the following:

Reviewer: ```suggestion
**With** model adaptivity, there is an iterative process, because a model may not be sufficiently accurate (given the current input).
The call to `solve(micro_sims_input, dt)` leads to the following logic:
```

```python
self._model_adaptivity_controller.initialise_solve()

active_sim_ids = None
if self._is_adaptivity_on:
    active_sim_ids = self._adaptivity_controller.get_active_sim_local_ids()
output = None

while self._model_adaptivity_controller.should_iterate():
    self._model_adaptivity_controller.switch_models(
        self._mesh_vertex_coords,
        self._t,
        micro_sims_input,
        output,
        self._micro_sims,
        active_sim_ids,
    )

>     output = solve_variant(micro_sims_input, dt)

Reviewer: Shouldn't this `solve_variant` call be of `self._micro_sims`?

        """

    def get_state(self):
        """
        Return the state of the micro simulation.
        """

    def output(self):
        """
        This function writes output of the micro simulation in some form.
        It will be called with frequency set by configuration option `simulation_params: micro_output_n`
        This function is *optional*.
        """
```

For this the default MicroSimulation still serves as the model interface, while the `(set)|(get)_state()` methods
are called to transfer internal model parameters from one to another.
The list of provided models is interpreted in decreasing fidelity order. In other words, the first one
is likely to be the full order model, while subsequent ones are ROMs.

```python
def switching_function(
    resolutions: np.ndarray,
    locations: np.ndarray,
    t: float,
    inputs: list[dict],
    prev_output: dict,
    active: np.ndarray,
) -> np.ndarray:
    """
    Switching interface function, use as reference

    Parameters
    ----------
    resolutions : np.array - shape(N,)
        Array with resolution information as get_sim_class_resolution would return for a sim obj.
    locations : np.array - shape(N,D)

>         Array with gaussian points for all sims. D is the mesh dimension.

Reviewer: I am not sure about having macro locations as part of the switching function inputs. I could imagine the switching being solely based on the macro- and micro-scale data available at a particular point.

Author: I thought it might be useful to enforce FOM at certain locations, where the user can already expect high fluctuations in input/output or where maximal accuracy should be guaranteed.

Reviewer: > I thought it might be useful to enforce FOM at certain locations, where the user can already expect high fluctuations in input/output or where maximal accuracy should be guaranteed.

Yes, that should be possible and required. But my comment was still about why the macro-scale coordinates need to be passed to the switching function. I look at the switching function as a function to be called individually at each location. If the switching mechanism requires a macro-scale location, it can be passed on a per-micro-scale simulation basis. I think having a single call-type switching mechanism is beneficial for the user.
